packages/server-core/src/user/login/login.class.ts

```diff
@@ -77,9 +79,63 @@ export class LoginService implements ServiceInterface {
       }
       if (new Date() > new Date(result.data[0].expiresAt)) {
         logger.info('Login Token has expired')
+        await this.app.service(loginTokenPath).remove(result.data[0].id)
         return { error: 'Login link has expired' }
       }
       const identityProvider = await this.app.service(identityProviderPath).get(result.data[0].identityProviderId)
+      let addToLogin = false
+      if (result.data[0].associateUserId && params!.query?.associate === 'true') {
+        await this.app.service(identityProviderPath).patch(identityProvider.id, {
+          userId: result.data[0].associateUserId
+        })
+        await this.app.service(userLoginPath).create({
+          userId: result.data[0].associateUserId as UserID,
+          userAgent: params!.headers!['user-agent'],
+          identityProviderId: identityProvider.id,
+          ipAddress: params!.forwarded?.ip || ''
+        })
+      }
+      if (params!.query?.associate != null) addToLogin = true
+      const logins = await this.app.service(userLoginPath).find({
+        query: {
+          userId: identityProvider.userId
+        }
+      })
+      //Email identity-providers are created as type email, so new vs. existing logins can't be discerned by
+      //whether the current identity-provider is a guest. We're using logins === 0 and !addToLogin as a proxy for
+      //a brand-new login with that email, which should trigger the auto-association, and that not being true
+      //will be seen as an email that is established and shouldn't have the auto-association trigger
+      if (identityProvider.type === 'email' && logins.total === 0 && !addToLogin) {
+        console.log('')
+        const existingIdentityProviders = await this.app.service(identityProviderPath).find({
+          query: {
+            $or: [
+              {
+                email: identityProvider.token
```

Comment:
just checking this is supposed to be .token and not .email - assuming that the token corresponds to an emal?

Reply:
Yes. With identity-providers, `token` is the unique identifier for that IP. For email magiclink, `token` is the email address. For an SSO, it's `<sso_provider_name>:::<sso_identifier>`, e.g. `github:::1234abcd`, if your unique GitHub ID were 1234abcd.

The `email` field on identity-providers is used to store the email address associated with an SSO provider, e.g. if I'd created my GH account with `kyle@email.com`, that would be stored in the `email` field of my GH identity-provider, while `token` would be that `github:::1234abcd` value.